src/common/components/markdownViewer/markdownViewerComponent.tsx

```diff
@@ -1,33 +1,110 @@
+  className?: string;
+  location?: any; // Router HOC injected.
+  options?: MdrOptions;
+  codeStyle?: MdrCodeStyle;
 }
 
-const getMarkDownChildren = (content: string): React.ReactNode => {
-  let childrenComponent: React.ReactNode = null;
+class MarkDownViewer extends React.Component<MarkDownViewerComponentProps, {}> {
+  constructor(props) {
+    super(props);
+  }
 
-  if (content) {
-    childrenComponent = compile(content).tree;
+  private CreateMdrInstance = () => {
+    const setupParams: MdrSetup = {
+      routerLocation: this.props.location ? this.props.location.pathname : '',
+      options: this.props.options,
+      codeStyle: this.props.codeStyle,
+    };
+    return CreateMarkdownRender(setupParams);
   }
 
-  return childrenComponent;
-};
+  private mdr: Mdr = this.CreateMdrInstance();
+  private scrollableContainerRef: HTMLElement = null;
+  private allowNotifyScroll: boolean = true;
+
+  private setScrollableContainerRef = (input) => { this.scrollableContainerRef = input; };
+
+  private markdownToHTML = () => {
+    return {
+      __html: this.mdr.render(this.props.content || ''),
+    };
+  }
+
+  private doScrollToSourceLine = (targetSourceLine) => {
+    const renderedElements = ReactDOM.findDOMNode(this).getElementsByClassName(SOURCE_LINE_CLASSNAME);
+    const scrollOffset = getPixelOffsetForSourceLine(renderedElements, targetSourceLine);
+    const componentPosition = this.scrollableContainerRef.getBoundingClientRect().top;
+    this.allowNotifyScroll = false;
+    this.scrollableContainerRef.scrollTop += scrollOffset - ((componentPosition > 0) ? componentPosition : 0);
+  }
+
+  private notifySourceLine = throttle(() => {
+    const componentPosition = this.scrollableContainerRef.getBoundingClientRect().top;
+    const renderedElements = ReactDOM.findDOMNode(this).getElementsByClassName(SOURCE_LINE_CLASSNAME);
+    const lineNum = getSourceLineForPixelOffset(renderedElements, componentPosition > 0 ? componentPosition : 0);
+    this.props.onScrollSourceLine(lineNum);
+  }, 25);
 
-export const MarkDownViewerComponent: React.StatelessComponent<MarkDownViewerComponentProps> = ({ content }) => {
-  return (
-    <div>
-      {getMarkDownChildren(content)}
-    </div>
-  );
+  private handleScroll = (event) => {
+    if (this.allowNotifyScroll) {
+      if (this.props.onScrollSourceLine) {
+        this.notifySourceLine();
+      }
+    } else {
+      this.allowNotifyScroll = true;
+    }
+  }
+
+  public shouldComponentUpdate(nextProps, nextState) {
+    if (nextProps.scrollSourceLine !== undefined && nextProps.scrollSourceLine !== this.props.scrollSourceLine) {
+      this.doScrollToSourceLine(nextProps.scrollSourceLine);
+      return false;
+    }
+    return true;
+  }
+
+  public componentWillUpdate(nextProps, nextState) {
+    // New render instance upon render options update.
+    if (nextProps.options !== this.props.options) {
+      this.mdr = this.CreateMdrInstance();
```

Comment:
When is it suppose that dynamically are going to change options? I think is not necessary if we have `markDownRender` instance in `State`

Reply:
Well that is because I thought it could be useful to expose markdonw-it options for parent components instead of the whole instance. My mind betrayed me trying to avoid dependency hell  :rofl:  ... too many years in C++ sorry. 
So yes, if you prefer to have the instance exposed in state, it makes no sense to have markdown-it options as component properties. But do we want the instance as public state or hidden as internal state?

Reply:
I would try to stick to state, whenever makes sense

Reply:
**Done**. It does not apply since the refactoring of [0b86b91](https://github.com/Lemoncode/LeanMood/pull/168/commits/0b86b91f18f426363c4ca3879acc90dd1c589771)